src/class-cache-collector.php

```diff
@@ -7,9 +7,517 @@
+								$type,
+								$key,
+								$cache_group,
+							),
+							[
+								'key'   => $key,
+								'group' => $cache_group,
+								'type'  => $type,
+							]
+						);
+					}
+				}
+			}
+		}
+
+		// Update the keys if any were removed.
+		if ( $dirty ) {
+			$this->store_keys( $keys );
+		}
+
+		return $this;
+	}
+
+	/**
+	 * Retrieve the storage name for the collector.
+	 *
+	 * This is the post name of the post that stores the cache collection keys
+	 * for a collection.
+	 *
+	 * @return string
+	 */
+	public function get_storage_name(): string {
+		return md5( $this->collection );
+	}
+
+	/**
+	 * Retrieve the parent object for the collector.
+	 *
+	 * The parent is the post where the cache keys will be stored for the
+	 * cache collection. If an existing parent is set it will return that
+	 * object. If no parent is set it will attempt to retrieve the parent
+	 * post if it exists.
+	 *
+	 * @param bool $create Whether to create the parent if it doesn't exist.
+	 * @return WP_Post|WP_Term
+	 */
+	public function get_parent_object( bool $create = true ): WP_Post|WP_Term|null {
+		if ( $this->parent ) {
+			return $this->parent;
+		}
+
+		$post = get_posts( // phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.get_posts_get_posts
+			[
+				'name'            => $this->get_storage_name(),
+				'no_found_rows'   => true,
+				'post_status'     => 'publish',
+				'post_type'       => static::POST_TYPE,
+				'posts_per_page'  => 1,
+				'suppress_filter' => false,
+			]
+		);
+
+		if ( ! empty( $post ) ) {
+			$this->parent = array_shift( $post );
+			return $this->parent;
+		}
+
+		if ( ! $create ) {
+			return null;
+		}
+
+		$post_id = wp_insert_post( // phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.wp_insert_post_wp_insert_post
+			[
+				'post_name'   => $this->get_storage_name(),
+				'post_status' => 'publish',
+				'post_title'  => $this->get_storage_name(),
+				'post_type'   => static::POST_TYPE,
+			]
+		);
+
+		if ( is_wp_error( $post_id ) ) {
+			if ( $this->logger ) {
+				$this->logger->error(
+					sprintf(
+						'Failed to create parent post for cache collection %s',
+						$this->get_storage_name(),
+					),
+					[
+						'error' => $post_id->get_error_message(),
+					]
+				);
+			}
+
+			return null;
+		}
+
+		$this->parent = get_post( $post_id );
+
+		// Store the collection for reference later.
+		update_post_meta( $post_id, '_collection', $this->collection );
+
+		return $this->parent;
+	}
+
+	/**
+	 * Store keys in the parent post/term.
+	 *
+	 * @param array $keys The keys to store.
+	 * @return void
+	 */
+	protected function store_keys( array $keys ) {
+		$this->get_parent_object();
+
+		if ( ! $this->parent ) {
+			return;
+		}
 
+		match ( $this->parent::class ) {
+			WP_Post::class => update_post_meta( $this->parent->ID, static::META_KEY, $keys ),
+			WP_Term::class => update_term_meta( $this->parent->term_id, static::META_KEY, $keys ),
```

Comment:
The use of a single array of keys is still, I think, subject to the race condition that you and Boynes discussed in Slack. However, using the metadata API allows for a possible workaround: for each key, store a separate meta row. That way, concurrent requests would write to their own rows, not the same row? I haven't played that all the way out in my head, so maybe it's a dead end.

Reply:
Oooo, i like this

Reply:
I've thought this through and I think I grasp the idea here but I don't know (right now) how to implement it. I split out ticket #14 to investigate it further.